clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -459,7 +467,412 @@ void SemaHLSL::handleResourceClassAttr(Decl *D, const ParsedAttr &AL) {
+static std::string getHLSLResourceTypeStr(Sema &S, Decl *TheDecl) {
+  VarDecl *TheVarDecl = dyn_cast<VarDecl>(TheDecl);
+  HLSLBufferDecl *CBufferOrTBuffer = dyn_cast<HLSLBufferDecl>(TheDecl);
+
+  if (TheVarDecl) {
+    QualType TheQualTy = TheVarDecl->getType();
+    PrintingPolicy PP = S.getPrintingPolicy();
+    return QualType::getAsString(TheQualTy.split(), PP);
+  } else {
+    return CBufferOrTBuffer->isCBuffer() ? "cbuffer" : "tbuffer";
+  }
+}
+
+static void DiagnoseHLSLRegisterAttribute(Sema &S, SourceLocation &ArgLoc,
+                                          Decl *TheDecl, RegisterType regType) {
+
+  // Samplers, UAVs, and SRVs are VarDecl types
+  VarDecl *TheVarDecl = dyn_cast<VarDecl>(TheDecl);
+  // Cbuffers and Tbuffers are HLSLBufferDecl types
+  HLSLBufferDecl *CBufferOrTBuffer = dyn_cast<HLSLBufferDecl>(TheDecl);
+
+  // exactly one of these two types should be set
+  assert(((TheVarDecl && !CBufferOrTBuffer) ||
+          (!TheVarDecl && CBufferOrTBuffer)) &&
+         "either TheVarDecl or CBufferOrTBuffer should be set");
+
+  RegisterBindingFlags Flags = HLSLFillRegisterBindingFlags(S, TheDecl);
+  assert((int)Flags.Other + (int)Flags.Resource + (int)Flags.Basic +
+                 (int)Flags.UDT ==
+             1 &&
+         "only one resource analysis result should be expected");
+
+  int regTypeNum = static_cast<int>(regType);
+
+  // first, if "other" is set, emit an error
+  if (Flags.Other) {
+    S.Diag(ArgLoc, diag::err_hlsl_binding_type_mismatch) << regTypeNum;
+    return;
+  }
+
+  // next, if multiple register annotations exist, check that none conflict.
+  ValidateMultipleRegisterAnnotations(S, TheDecl, regType);
+
+  // next, if resource is set, make sure the register type in the register
+  // annotation is compatible with the variable's resource type.
+  if (Flags.Resource) {
+    const HLSLResourceClassAttr *resClassAttr = nullptr;
+    if (CBufferOrTBuffer) {
+      resClassAttr = CBufferOrTBuffer->getAttr<HLSLResourceClassAttr>();
+    } else if (TheVarDecl) {
+      resClassAttr =
+          getSpecifiedHLSLAttrFromVarDeclOrRecordDecl<HLSLResourceClassAttr>(
+              TheVarDecl, nullptr);
+    }
+
+    assert(resClassAttr &&
+           "any decl that set the resource flag on analysis should "
+           "have a resource class attribute attached.");
+    const llvm::hlsl::ResourceClass DeclResourceClass =
+        resClassAttr->getResourceClass();
+
+    // confirm that the register type is bound to its expected resource class
+    static RegisterType ExpectedRegisterTypesForResourceClass[] = {
+        RegisterType::SRV,
+        RegisterType::UAV,
+        RegisterType::CBuffer,
+        RegisterType::Sampler,
+    };
+    assert((int)DeclResourceClass <
+               std::size(ExpectedRegisterTypesForResourceClass) &&
+           "DeclResourceClass has unexpected value");
+
+    RegisterType ExpectedRegisterType =
+        ExpectedRegisterTypesForResourceClass[(int)DeclResourceClass];
```

Comment:
I think this conversion could be simplified without the static array. How about declaring RegisterType with the values from RegisterClass?

```
enum class RegisterType { 
  SRV = static_cast<int>(llvm::hlsl::ResourceClass::SRV),
  UAV = static_cast<int>(llvm::hlsl::ResourceClass::UAV),
  CBuffer = static_cast<int>(llvm::hlsl::ResourceClass::CBuffer),
  Sampler = static_cast<int>(llvm::hlsl::ResourceClass::Sampler)
};

static RegisterType ResourceClassToRegisterType(llvm::hlsl::ResourceClass RC) {
  assert(static_cast<int>(RC) >= static_cast<int>(RegisterType::SRV) &&
         static_cast<int>(RC) <= static_cast<int>(RegisterType::Sampler) &&
           "unexpected ResourceClass value");
  return static_cast<RegisterType>(static_cast<int>(RC));
}
```

Reply:
In my opinion it looks a little more complicated.

Reply:
That might be, but the actual code executed for the conversion is a no-op and there is no static array to allocate and initialize.